lmsweb/views.py

```diff
@@ -147,31 +154,79 @@ def exercises_page():
+    if user is None:
+        # should never happen, we checked session_id == solver_id
+        return fail(404, 'No such user')
+
+    kind = request.form.get('kind', '')
+    if (not kind) or (kind not in ('id', 'text')):
+        return fail(400, "Invalid kind")
+
+    try:
+        line_number = int(request.form.get('line', 0))
+    except ValueError:
+        line_number = 0
+    if line_number <= 0:
+        return fail(422, f"Invalid line number: {line_number}")
+
+    if kind == 'id':
+        comment_id = int(request.form.get('comment', 0))
+
+    elif kind == 'text':
+        text = request.form.get('comment', '')
+        if not text:
+            return fail(422, 'Empty comments are not allowed')
+            # TODO(LOW): Check if line number > MAX_SOLUTION_LINE_NUMBER
+        comment_id = CommentText.create(text=text).id
+    else:
+        # should never happend, kind was checked before
+        return fail(400, "Invalid kind")
+
+    Comment.create(
+        commenter=user,
+        timestamp=datetime.now(),
+        line_number=line_number,
+        comment=comment_id,
+        solution=solution
+    )
+    return jsonify({"success": "true"})
+
+
 @webapp.route('/comments', methods=['GET', 'POST'])
 @login_required
 def comment():
-    is_manager = session['role'] in HIGH_ROLES
-    if is_manager and request.method == 'POST':
-        solutionId = request.form['solutionId']
-        print(solutionId)
-        return jsonify('{"success": "true"}')
-
-    if request.method != 'GET':
-        return abort(405, "Must be GET or POST")
-
-    solution_id = int(request.args['solutionId'])
-    solution = Solution.select(Solution).where(Solution.id == solution_id)
-    if is_manager or solution.solver == session['id']:
-        if request.args.get('act') == 'fetch':
-            return jsonify(Comment.by_solution(solution_id))
-        if request.args.get('act') == 'delete':
-            comment_id = int(request.args.get('commentId'))
-            # TODO: Handle if not found
-            Comment.get(
-                Comment.comment == comment_id,
-                Comment.solution == solution_id,
-            ).delete_instance()
-            return jsonify({"success": "true"})
+    act = request.args.get('act')
+
+    solution_id = int(request.form.get('solutionId', 0))
+    session_id = int(session['id'])
+
+    solution = Solution.get_or_none(Solution.id == solution_id)
+    if solution is None:
+        return fail(404, f"No such solution {solution_id}")
+
+    solver_id = solution.solver.id
+    if solver_id != session_id:
+        return fail(401, "You aren't allowed to watch this page.")
+
+    if act == 'fetch':
+        return jsonify(Comment.by_solution(solution_id))
+
+    if session['role'] in HIGH_ROLES:
+        return fail(401, "You must be an admin to view this page.")
+
+    if act == 'delete':
+        comment_id = int(request.args.get('commentId'))
+        comment_ = Comment.get_or_none(Comment.id == comment_id)
```

Comment:
just `comment` will do actually

Reply:
shadowing the `comment` function name. that why i used _